### PR TITLE
Wait for transaction commit before pushing tasks

### PR DESCRIPTION
**Pull request by erikh360**
We were experiencing issues with our USSD service, our sentry was showing these errors:

```
IntegrityError: insert or update on table "channels_channellog" violates foreign key constraint "channels_channellog_msg_id_e40e6612_fk_msgs_msg_id"
DETAIL:  Key (msg_id)=(260040) is not present in table "msgs_msg".
```

After investigating I found the message object would always exist in the DB, but the queue started processing the message before the original transaction was committed. It will now wait for the transaction to commit before placing things on the queue.

We have this fix running in one of our Production environments and the error has disappeared.

### CODE REVIEW

**Comment by erikh360** [comment]
@nicpottier Please review, this is the fix for the discussion we had titled `USSD timing out and/or ending prematurely`, thanks.

File: temba/settings.py.dev [hunk 1]
```diff
@@ -51,6 +51,18 @@ DATABASES = {
...
+        'ENGINE': 'django.contrib.gis.db.backends.postgis',
+        'NAME': 'temba',
+        'USER': 'temba',
+        'PASSWORD': 'temba',
+        'HOST': 'localhost',
+        'PORT': '',
+        'ATOMIC_REQUESTS': True,
+        'CONN_MAX_AGE': 60,
+        'OPTIONS': {
+        }
```
Comment (nicpottier) [inline temba/settings.py.dev]: What is this addition for?

File: temba/msgs/tests.py [hunk 2]
```diff
@@ -2103,3 +2107,120 @@ def test_render(self):
         # exception if tag not used correctly
         self.assertRaises(ValueError, self.render_template, '{% load sms %}{% render with bob %}{% endrender %}')
         self.assertRaises(ValueError, self.render_template, '{% load sms %}{% render as %}{% endrender %}')
+
+
+class CeleryTaskTest(TransactionTestCase):
```
Comment (nicpottier) [inline temba/msgs/tests.py]: I think you could avoid a lot of the boilerplate of setting up an org etc, if you just subclass `TembaTest`

File: temba/msgs/models.py [hunk 3]
```diff
@@ -789,7 +791,15 @@ def send_messages(cls, all_msgs):
...
             if task_priority is None:
                 task_priority = DEFAULT_PRIORITY
-            push_task(task_msgs[0]['org'], MSG_QUEUE, SEND_MSG_TASK, task_msgs, priority=task_priority)
+
+            tasks_to_push.append((task_msgs, task_priority))
+
+        on_transaction_commit(lambda: cls.push_tasks(tasks_to_push))
+
+    @classmethod
+    def push_tasks(cls, tasks):
```
Comment (nicpottier) [inline temba/msgs/models.py]: I don't think there's any real need for this, it is basically just recreating what `on_transaction_commit` already does. (it has an internal list, so you can call `on_transaction_commit` more than once)

Just changing all the push_task to `on_transaction_commit` should work no? (being careful about shared variables)

File: temba/settings.py.dev (hunk 1 above)
Comment (erikh360) [inline temba/settings.py.dev]: I use it in the tests inside the `assertInDB` function, this opens a separate DB transaction to make sure the record is actually committed.

File: temba/msgs/tests.py (hunk 2 above)
Comment (erikh360) [inline temba/msgs/tests.py]: I had to use the `TransactionTestCase` to get the DB transaction tests working. It's the only way I could get the original code to fail with the tests.

**Comment by erikh360** [comment]
@nicpottier Ready for another look!

File: temba/msgs/tests.py (hunk 2 above)
Comment (nicpottier) [inline temba/msgs/tests.py]: Hrmm.. any chance you could use insane Python multiple inheritance to get the best of both? Looking at `TransactionTestCase` I don't think its magic overlaps much with what is provided by TembaTest so I don't think they would walk over each other.

Sorry to put this back in your court but man our tests are already so full of repetition already we are trying not to have that debt grow even more.

Comment (erikh360) [inline temba/msgs/tests.py]: I'll see what I can do. 🙂

Comment (erikh360) [inline temba/msgs/tests.py]: @nicpottier Please have a look now. `TembaTest` actually inherits from `TransactionTestCase` a few levels up, so I just had to disable the rollback things.

File: temba/msgs/tests.py [hunk 4]
```diff
@@ -2103,3 +2109,84 @@ def test_render(self):
...
+        try:
+            type(obj).objects.using('default2').get(pk=obj.pk)
+        except obj.DoesNotExist:
+            self.fail(fullmsg)
+
+    def test_reply_task_added(self):
+        send_task_orig = current_app.send_task
+        orig_push_task = models.push_task
+
+        settings.CELERY_ALWAYS_EAGER = False
```
Comment (rowanseymour) [inline temba/msgs/tests.py]: Use @override_settings decorator on the test method so these automatically get reset

File: temba/msgs/tests.py [hunk 5]
```diff
@@ -2103,3 +2109,84 @@ def test_render(self):
...
+        settings.SEND_MESSAGES = True
+
+        try:
+            def new_send_task(name, args=(), kwargs={}, **opts):  # pragma: needs cover
+                self.handle_push_task(name)
+
+            # monkey patch the regular send_task with our method
+            current_app.send_task = new_send_task
```
Comment (rowanseymour) [inline temba/msgs/tests.py]: Use `with patch(...):`

**Comment by erikh360** [comment]
@rowanseymour Please have a look again.

**Comment by rowanseymour** [comment]
Thanks, will be good to get this deployed!

**Pull request merged**